File: libraries/dotnet/UrcConverter/Sources/Osu/Parse.fs
namespace UrcConverter.Sources.Osu

module Parse =

    open System
    open System.Globalization
    open FParsec
    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources
    open UrcConverter.Sources.Osu.Model

    type private State = { Beatmap: Beatmap; Section: string }

    let private pNumber: Parser<float, State> =
        let invariant (token: string) =
            Double.Parse(token, NumberStyles.Float, CultureInfo.InvariantCulture)

        pipe2
            (opt (pchar '-'))
            (many1 digit .>>. opt (pchar '.' >>. many1 digit))
            (fun sign (whole, fraction) ->
                let magnitude =
                    match fraction with
                    | None -> invariant (String(Array.ofList whole))
                    | Some digits ->
                        invariant $"{String(Array.ofList whole)}.{String(Array.ofList digits)}"

                if sign.IsSome then -magnitude else magnitude)

    let private pNumberW: Parser<float, State> =
        many (pchar ' ') >>. pNumber .>> many (pchar ' ')

    let private lineText: Parser<string, State> =
        manySatisfy (fun c -> c <> '\r' && c <> '\n')

    let private numberField
        (label: string)
        (value: string)
        (update: Beatmap -> float -> Beatmap)
        : Parser<unit, State> =
        match runParserOnString (pNumber .>> eof) Unchecked.defaultof<State> "" value with
        | Success(number, _, _) ->
            updateUserState (fun state ->
                { state with
                    Beatmap = update state.Beatmap number
                })
        | Failure _ -> fail $"invalid {label}: '{value}'"

    let private pKeyValue: Parser<string * string, State> =
        pipe2
            (many1Satisfy (fun c -> c <> ':' && c <> '\r' && c <> '\n') .>> pchar ':')
            lineText
            (fun key value -> key.Trim(), value.Trim())

    let private applyKeyValue ((key, value): string * string) : Parser<unit, State> =
        match key with
        | "Mode" ->
            numberField "Mode" value (fun beatmap number ->
                { beatmap with Mode = Shared.roundMs number })
        | "CircleSize" ->
            numberField "CircleSize" value (fun beatmap number ->
                { beatmap with
                    CircleSize = Some number
                })
        | "OverallDifficulty" ->
            numberField "OverallDifficulty" value (fun beatmap number ->
                { beatmap with
                    OverallDifficulty = Some number
                })
        | "Title"
        | "TitleUnicode"
        | "Artist"
        | "ArtistUnicode"
        | "Creator"
        | "Version" ->
            updateUserState (fun state ->
                let beatmap = state.Beatmap

                let beatmap =
                    match key with
                    | "Title" -> { beatmap with Title = Some value }
                    | "TitleUnicode" ->
                        { beatmap with
                            TitleUnicode = Some value
                        }
                    | "Artist" -> { beatmap with Artist = Some value }
                    | "ArtistUnicode" ->
                        { beatmap with
                            ArtistUnicode = Some value
                        }
                    | "Creator" -> { beatmap with Creator = Some value }
                    | _ -> { beatmap with Version = Some value }

                { state with Beatmap = beatmap })
        | _ -> preturn ()

    let private pTimingBody: Parser<unit, State> =
        sepBy pNumberW (pchar ',')
        >>= fun fields ->
            match fields with
            | [] -> preturn ()
            | [ _ ] -> fail "timing point needs at least 2 fields"
            | time :: beatLength :: _ ->
                let meter = List.tryItem 2 fields |> Option.map Shared.roundMs |> Option.defaultValue 4

                let uninherited =
                    List.tryItem 6 fields
                    |> Option.map (fun value -> Shared.roundMs value <> 0)
                    |> Option.defaultValue true

                updateUserState (fun state ->
                    let point =
                        {
                            Time = Shared.roundMs time
                            BeatLength = beatLength
                            Meter = meter
                            Uninherited = uninherited
                        }

                    { state with
                        Beatmap =
                            { state.Beatmap with
                                TimingPoints = point :: state.Beatmap.TimingPoints
                            }
                    })

    let private pHitObjectBody: Parser<unit, State> =
        pipe4
            (pNumberW .>> pchar ',')
            (pNumberW .>> pchar ',')
            (pNumberW .>> pchar ',')
            pNumberW
            (fun x _ time typeValue ->
                Shared.roundMs x, Shared.roundMs time, Shared.roundMs typeValue)
        >>= fun (x, time, typeBits) ->
            let isHold = typeBits &&& 128 <> 0

            if not isHold && typeBits &&& 1 = 0 then
                fail $"unsupported hit object type: {typeBits}"
            else
                let endTimeParser =
                    if isHold then
                        (pchar ',' >>. pNumberW >>. pchar ',' >>. pNumberW)
                        |>> fun endValue -> Some(Shared.roundMs endValue)
                    else
                        preturn None

                endTimeParser
                >>= fun endValue ->
                    updateUserState (fun state ->
                        let hitObject =
                            {
                                X = x
                                Time = time
                                IsHold = isHold
                                EndTime = endValue
                            }

                        { state with
                            Beatmap =
                                { state.Beatmap with
                                    HitObjects = hitObject :: state.Beatmap.HitObjects
                                }
                        })

    let private pSectionHeader: Parser<unit, State> =
        pchar '[' >>. manySatisfy (fun c -> c <> ']') .>> pchar ']'
        >>= fun name -> updateUserState (fun state -> { state with Section = name.Trim() })

    let private pLine: Parser<unit, State> =
        getUserState
        >>= fun state ->
            let body =
                match state.Section with
                | "General"
                | "Metadata"
                | "Difficulty" ->
                    // A line without a colon is skipped, matching the original
                    // parser; a bad value in a known field still fails because
                    // `attempt` only wraps the key-value shape.
                    (attempt pKeyValue >>= applyKeyValue) <|> preturn ()
                | "TimingPoints" -> pTimingBody
                | "HitObjects" -> pHitObjectBody
                | _ -> preturn ()

            many (pchar ' ')
            >>. choice [ attempt pSectionHeader; attempt (pstring "//") |>> ignore; body ]
            .>> (lineText |>> ignore)
            .>> newline

    let private pOsuFile: Parser<Beatmap, State> =
        many pLine .>> eof >>. getUserState
        |>> fun state ->
            let beatmap = state.Beatmap

            { beatmap with
                TimingPoints = List.rev beatmap.TimingPoints
                HitObjects = List.rev beatmap.HitObjects
            }

    let parse (text: string) : Result<Beatmap, UrcError> =
        let stripped =
            if text.StartsWith("\uFEFF", StringComparison.Ordinal) then
                text[1..]
            else
                text

        let normalized =
            if stripped.EndsWith("\n") then
                stripped
            else
                stripped + "\n"

        let initial =
            {
                Beatmap =
                    {
                        Mode = 3
                        Title = None
                        TitleUnicode = None
                        Artist = None
                        ArtistUnicode = None
                        Creator = None
                        Version = None
                        CircleSize = None
                        OverallDifficulty = None
                        TimingPoints = []
                        HitObjects = []
                    }
                Section = ""
            }

        match runParserOnString pOsuFile initial "" normalized with
        | Success(beatmap, _, _) -> Result.Ok beatmap
        | Failure(_, parseError, _) ->
            Result.Error(
                UrcError.Syntax(int parseError.Position.Line + 1, $"invalid .osu: {parseError}")
            )

File: libraries/python/src/urc_converter/sources/osu/parse.py
"""Parser for .osu text."""

from ...error import UrcError
from .model import OsuBeatmap, OsuHitObject, OsuTimingPoint


def parse_osu(text: str) -> OsuBeatmap:
	"""Parse .osu text into a source model."""
	beatmap = OsuBeatmap()
	section = None

	for line_no, raw in enumerate(_lines(text), start=1):
		line = raw.strip()

		if line.startswith("[") and line.endswith("]"):
			section = line[1:-1]
			continue

		if line == "" or line.startswith("//") or section is None:
			continue

		if section == "General":
			_general(beatmap, line, line_no)
		elif section == "Metadata":
			_metadata(beatmap, line)
		elif section == "Difficulty":
			_difficulty(beatmap, line, line_no)
		elif section == "TimingPoints":
			beatmap.timing_points.append(_timing_point(line, line_no))
		elif section == "HitObjects":
			beatmap.hit_objects.append(_hit_object(line, line_no))

	return beatmap


def _lines(text: str) -> list[str]:
	if text.startswith("\ufeff"):
		text = text[1:]
	return text.replace("\r\n", "\n").replace("\r", "\n").split("\n")


def _general(beatmap: OsuBeatmap, line: str, line_no: int) -> None:
	key, _, value = line.partition(":")
	if key.strip() == "Mode":
		beatmap.mode = _to_int(value.strip(), line_no, "Mode")


def _metadata(beatmap: OsuBeatmap, line: str) -> None:
	key, _, value = line.partition(":")
	attribute = {
		"Title": "title",
		"TitleUnicode": "title_unicode",
		"Artist": "artist",
		"ArtistUnicode": "artist_unicode",
		"Creator": "creator",
		"Version": "version",
	}.get(key.strip())
	if attribute is not None:
		setattr(beatmap, attribute, value.strip())


def _difficulty(beatmap: OsuBeatmap, line: str, line_no: int) -> None:
	key, _, value = line.partition(":")
	name = key.strip()

	if name == "CircleSize":
		beatmap.circle_size = _to_float(value.strip(), line_no, "CircleSize")
	elif name == "OverallDifficulty":
		beatmap.overall_difficulty = _to_float(value.strip(), line_no, "OverallDifficulty")


def _timing_point(line: str, line_no: int) -> OsuTimingPoint:
	fields = [field.strip() for field in line.split(",")]
	if len(fields) < 2:
		raise UrcError("syntax", line_no, f"timing point needs at least 2 fields: {line!r}")

	return OsuTimingPoint(
		time=_to_int(fields[0], line_no, "timing time"),
		beat_length=_to_float(fields[1], line_no, "beat length"),
		meter=_to_int(fields[2], line_no, "meter") if len(fields) > 2 and fields[2] else 4,
		uninherited=bool(int(fields[6])) if len(fields) > 6 and fields[6] else True,
	)


def _hit_object(line: str, line_no: int) -> OsuHitObject:
	fields = [field.strip() for field in line.split(",")]
	if len(fields) < 5:
		raise UrcError("syntax", line_no, f"hit object needs at least 5 fields: {line!r}")

	x = _to_int(fields[0], line_no, "hit object x")
	time = _to_int(fields[2], line_no, "hit object time")
	type_bits = _to_int(fields[3], line_no, "hit object type")

	is_hold = bool(type_bits & 128)
	if not is_hold and not type_bits & 1:
		raise UrcError("syntax", line_no, f"unsupported hit object type: {type_bits}")

	end_time = None
	if is_hold:
		if len(fields) < 6:
			raise UrcError("syntax", line_no, f"hold note needs an end time: {line!r}")
		end_time = _to_int(fields[5].split(":", 1)[0], line_no, "hold end time")

	return OsuHitObject(x=x, time=time, is_hold=is_hold, end_time=end_time)


def _to_int(token: str, line_no: int, label: str) -> int:
	try:
		return round(float(token))
	except ValueError:
		raise UrcError("syntax", line_no, f"invalid {label}: {token!r}") from None


def _to_float(token: str, line_no: int, label: str) -> float:
	try:
		return float(token)
	except ValueError:
		raise UrcError("syntax", line_no, f"invalid {label}: {token!r}") from None
